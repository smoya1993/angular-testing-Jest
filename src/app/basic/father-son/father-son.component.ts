import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { Client } from '../interfaces';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styleUrls: ['./father-son.component.scss']
})
export class FatherSonComponent implements OnInit {

  constructor() { }
  @Input() client?: Client;
  @Output() onDeleteClient  = new EventEmitter();
  @Output() onClientUpdated = new EventEmitter<Client>();

  onDelete() {
    this.client = undefined;
    this.onDeleteClient.emit();
  }

  onChange( id: number ) {
    if ( !this.client ) return;
    
    // this.client.id = newId; // no usar
    this.client = { ...this.client, id };

    this.onClientUpdated.emit({ ...this.client })

  }

  ngOnInit(): void {
  }

}
