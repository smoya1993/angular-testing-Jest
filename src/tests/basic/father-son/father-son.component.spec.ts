import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FatherSonComponent } from '../../../app/basic/father-son/father-son.component';

describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ FatherSonComponent ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
    jest.clearAllMocks();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should have to match snapshot', () => {
    expect(compiled.innerHTML).toMatchSnapshot();    
  });

  it('should not show any botons if client is empty/null', () => {
    const buttons = compiled.querySelectorAll('button');
    expect( buttons.length ).toBe(0);
  });

  it('should show 2 botons if client is OK', () => { 
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    const buttons = compiled.querySelectorAll('button');
    expect( buttons.length ).toBe(2);
  });

  it('should have to match snapshot if client is OK', () => { 
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();
    expect( compiled ).toMatchSnapshot();
  });

  test('should have to emit onDeleteClient with delete buttom', () => {

    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    jest.spyOn( component.onDeleteClient, 'emit' );
    
    const btnDelete = compiled.querySelector('[data-test=btn-delete]');
    btnDelete?.dispatchEvent( new Event('click') );

    expect( component.onDeleteClient.emit ).toHaveBeenCalled();
    
  });

  test('should have to emit onClientUpdated with Cambiar ID buttom', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();
    jest.spyOn( component.onClientUpdated, 'emit' );
    const btnChangeId = compiled.querySelector('[data-test=btn-id]');
    btnChangeId?.dispatchEvent( new Event('click') );
    expect( component.onClientUpdated.emit ).toHaveBeenCalledWith({
      id: 5,
      name: 'Juan'
    }); 
  });

  test('should have to emit onChangeClient with specific ID if client is OK', () => {

    jest.spyOn( component.onClientUpdated, 'emit' );
    component.onChange(10);
    expect( component.onClientUpdated.emit ).not.toHaveBeenCalled();

    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();
    component.onChange(10);

    expect( component.onClientUpdated.emit ).toHaveBeenCalledWith({
      id: 10,
      name: 'Juan'
    });

  });

});
