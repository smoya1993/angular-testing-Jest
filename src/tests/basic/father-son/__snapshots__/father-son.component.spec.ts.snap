// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent should have to match snapshot 1`] = `"<h5>Hijo</h5><code></code><br><br><!--bindings={}--><!--bindings={}-->"`;

exports[`FatherSonComponent should have to match snapshot if client is OK 1`] = `
<div
  id="root4"
>
  <h5>
    Hijo
  </h5>
  <code>
    {
  "id": 1,
  "name": "Juan"
}
  </code>
  <br />
  <br />
  <button
    class="btn btn-outline-primary"
    data-test="btn-id"
  >
     Cambiar ID

  </button>
  
  <button
    class="btn btn-outline-danger"
    data-test="btn-delete"
  >
     Borrar cliente

  </button>
  
</div>
`;
